src/github.com/couchbase/sync_gateway/db/kv_change_index.go

```diff
@@ -295,137 +332,25 @@ func (k *kvChangeIndex) EnableChannelIndexing(enable bool) {
 // The JSON must be the raw document from the bucket, with the metadata and all.
 func (k *kvChangeIndex) DocChanged(docID string, docJSON []byte, seq uint64, vbNo uint16) {
 
-	entryTime := time.Now()
-	// ** This method does not directly access any state of c, so it doesn't lock.
-
-	// TODO: in order to work around https://github.com/couchbase/sync_gateway/issues/1139,
-	// I had to remove the code which spawned a goroutine, since it broke the strict
-	// ordering of how changes were being processed.  (see issue 1139).  To achieve
-	// parallelism, this kvChangeIndex could maintain a pool of worker goroutines
-	// that are reading from a channel.
-
-	// Is this a user/role doc?
-	if strings.HasPrefix(docID, auth.UserKeyPrefix) {
-		k.processPrincipalDoc(docID, docJSON, true, vbNo, seq)
-		return
-	} else if strings.HasPrefix(docID, auth.RoleKeyPrefix) {
-		k.processPrincipalDoc(docID, docJSON, false, vbNo, seq)
-		return
-	}
-
-	// First unmarshal the doc (just its metadata, to save time/memory):
-	doc, err := unmarshalDocumentSyncData(docJSON, false)
-	if err != nil || !doc.hasValidSyncData(false) {
-		base.Warn("ChangeCache: Error unmarshaling doc %q: %v", docID, err)
-		return
-	}
-
-	// Record a histogram of the  feed's lag:
-	feedLag := time.Since(doc.TimeSaved) - time.Since(entryTime)
-	lagMs := int(feedLag/(100*time.Millisecond)) * 100
-	changeCacheExpvars.Add(fmt.Sprintf("lag-feed-%04dms", lagMs), 1)
-
-	// Now add the entry for the new doc revision:
-	change := &LogEntry{
-		Sequence:     seq,
-		DocID:        docID,
-		RevID:        doc.CurrentRev,
-		Flags:        doc.Flags,
-		TimeReceived: time.Now(),
-		TimeSaved:    doc.TimeSaved,
-		Channels:     doc.Channels,
-		VbNo:         uint16(vbNo),
-	}
-	base.LogTo("DIndex+", "Received #%d after %3dms (%q / %q)", change.Sequence, int(feedLag/time.Millisecond), change.DocID, change.RevID)
-
-	if change.DocID == "" {
-		base.Warn("Unexpected change with empty DocID for sequence %d, vbno:%d", doc.Sequence, vbNo)
-		changeCacheExpvars.Add("changes_without_id", 1)
-		return
+	// Incoming docs are assigned to the appropriate unmarshalWorker for the vbucket, in order
+	// to ensure docs are processed in sequence for a given vbucket.
+	unmarshalWorker := k.unmarshalWorkers[vbNo]
+	if unmarshalWorker == nil {
+		// Initialize new worker that sends results to k.pending for processing by the indexPending loop
+		unmarshalWorker = NewUnmarshalWorker(k.pending)
+		k.unmarshalWorkers[vbNo] = unmarshalWorker
```

Comment:
DocChanged shouldn't ever be called by multiple goroutines  - it's doing sequential processing of entries as they show up on the DCP feed.

I agree it would be nice to future-proof DocChanged, but since it's a potential bottleneck (every entry coming over DCP blocks until DocChanged returns), I wanted to avoid the overhead of a mutex check.

Maybe just a comment warning about DocChanged not being safe for concurrent calls?